real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java

```diff
@@ -32,75 +49,219 @@ public void updatePositionEstimate(GPSPoseMessage m) {
+                double dx = LocalizerUtil.convertLonToMeters(wayPoints.get(i).toGpsPoseMessage(0).getLongitude()) - LocalizerUtil.convertLonToMeters
+                        (wayPoints.get(closestIndex).getLongitude());
+                double dy = LocalizerUtil.convertLatToMeters(wayPoints.get(i).toGpsPoseMessage(0).getLatitude()) - LocalizerUtil.convertLatToMeters
+                        (wayPoints.get(closestIndex).getLatitude());
+
+                double ch = pose.getCurrentState().get(3, 0);
+                double dh = Math.atan2(dy, dx);
+
+                double theta = dh-ch;
+                // pick the first point that is at least lookahead away
+                if (GPSPoseMessage.getDistance(pose, wayPoints.get(i).toGpsPoseMessage(0)) * Math.cos(theta) > lookahead) {
+                    target = wayPoints.get(i);
+                    currentWaypoint = target;
+                    break;
+                }
+            }
         }
 
+        // steer towards it
+        double currentOrientation = pose.getCurrentState().get(3, 0);
+        double dx = LocalizerUtil.convertLonToMeters(target.getLongitude()) - LocalizerUtil.convertLonToMeters(pose.getLongitude());
+        double dy = LocalizerUtil.convertLatToMeters(target.getLatitude()) - LocalizerUtil.convertLatToMeters(pose.getLatitude());
+        double desiredHeading = Math.atan2(dy, dx);
+        desiredHeading = Util.normalizeAngleRad(desiredHeading);
 
-        double delta = 10; //meters
-        //pick the first point that is at least delta away
-        //pick the point to follow
-        int targetIndex = closestIndex;
-        while (targetIndex < wayPoints.size() && GPSPoseMessage.getDistance(pose, wayPoints.get(targetIndex).toGpsPoseMessage(0)) < delta) {
-            targetIndex = targetIndex + 1;
+        double alpha = desiredHeading - currentOrientation;
+        double L = RobobuggyKFLocalizer.WHEELBASE_IN_METERS;
+        double ld = GPSPoseMessage.getDistance(target.toGpsPoseMessage(0), pose) + L / 2;
+        return Math.atan2(2*L*Math.sin(alpha), ld);
+    }
+
+    private double purePursuitController() {
+        // https://www.ri.cmu.edu/pub_files/2009/2/Automatic_Steering_Methods_for_Autonomous_Automobile_Path_Tracking.pdf
+        // section 2.2
+
+        int closestIndex = getClosestIndex(wayPoints, pose);
+
+        double K = 2.5;
+        double velocity = pose.getCurrentState().get(2, 0);
+        double lookaheadLowerBound = 5.0;
+        double lookaheadUpperBound = 25.0;
+        double lookahead = K * velocity;
+        if(lookahead < lookaheadLowerBound) {
+            lookahead = lookaheadLowerBound;
+        }
+        else if(lookahead > lookaheadUpperBound) {
+            lookahead = lookaheadUpperBound;
+        }
+
+        //pick the first point that is at least lookahead away, then point buggy toward it
+        int lookaheadIndex = 0;
+        for(lookaheadIndex = closestIndex; lookaheadIndex < wayPoints.size(); lookaheadIndex++) {
```

Comment:
Maybe break up this method by having a different method to find the lookahead index